src/compiler/scanner.ts

```diff
@@ -301,6 +301,120 @@ const regExpFlagToFirstAvailableLanguageVersion = new Map<RegularExpressionFlags
+
+        // Simple Single-Character Tokens
+        [CharacterCodes.openParen, TokenCategory.SimpleToken | SyntaxKind.OpenParenToken],
+        [CharacterCodes.closeParen, TokenCategory.SimpleToken | SyntaxKind.CloseParenToken],
+        [CharacterCodes.comma, TokenCategory.SimpleToken | SyntaxKind.CommaToken],
+        [CharacterCodes.colon, TokenCategory.SimpleToken | SyntaxKind.ColonToken],
+        [CharacterCodes.semicolon, TokenCategory.SimpleToken | SyntaxKind.SemicolonToken],
+        [CharacterCodes.openBracket, TokenCategory.SimpleToken | SyntaxKind.OpenBracketToken],
+        [CharacterCodes.closeBracket, TokenCategory.SimpleToken | SyntaxKind.CloseBracketToken],
+        [CharacterCodes.openBrace, TokenCategory.SimpleToken | SyntaxKind.OpenBraceToken],
+        [CharacterCodes.closeBrace, TokenCategory.SimpleToken | SyntaxKind.CloseBraceToken],
+        [CharacterCodes.tilde, TokenCategory.SimpleToken | SyntaxKind.TildeToken],
+        [CharacterCodes.at, TokenCategory.SimpleToken | SyntaxKind.AtToken],
+
+        // Digits
+        [CharacterCodes._0, TokenCategory.Digit],
+        [CharacterCodes._1, TokenCategory.Digit],
+        [CharacterCodes._2, TokenCategory.Digit],
+        [CharacterCodes._3, TokenCategory.Digit],
+        [CharacterCodes._4, TokenCategory.Digit],
+        [CharacterCodes._5, TokenCategory.Digit],
+        [CharacterCodes._6, TokenCategory.Digit],
+        [CharacterCodes._7, TokenCategory.Digit],
+        [CharacterCodes._8, TokenCategory.Digit],
+        [CharacterCodes._9, TokenCategory.Digit],
+
+        [CharacterCodes.exclamation, TokenCategory.RecognizedMisc],
+        [CharacterCodes.doubleQuote, TokenCategory.RecognizedMisc],
+        [CharacterCodes.singleQuote, TokenCategory.RecognizedMisc],
+        [CharacterCodes.backtick, TokenCategory.RecognizedMisc],
+        [CharacterCodes.percent, TokenCategory.RecognizedMisc],
+        [CharacterCodes.ampersand, TokenCategory.RecognizedMisc],
+        [CharacterCodes.asterisk, TokenCategory.RecognizedMisc],
+        [CharacterCodes.plus, TokenCategory.RecognizedMisc],
+        [CharacterCodes.minus, TokenCategory.RecognizedMisc],
+        [CharacterCodes.dot, TokenCategory.RecognizedMisc],
+        [CharacterCodes.slash, TokenCategory.RecognizedMisc],
+        [CharacterCodes.lessThan, TokenCategory.RecognizedMisc],
+        [CharacterCodes.equals, TokenCategory.RecognizedMisc],
+        [CharacterCodes.greaterThan, TokenCategory.RecognizedMisc],
+        [CharacterCodes.question, TokenCategory.RecognizedMisc],
+        [CharacterCodes.caret, TokenCategory.RecognizedMisc],
+        [CharacterCodes.bar, TokenCategory.RecognizedMisc],
+        [CharacterCodes.backslash, TokenCategory.RecognizedMisc],
+        [CharacterCodes.hash, TokenCategory.RecognizedMisc],
+        [CharacterCodes.replacementCharacter, TokenCategory.RecognizedMisc],
+    ]
+) {
+    if (key < tokenCategoryLookup.length) {

```

Comment:
Can you be more specific? In the first two commits, I split this out from a map to an array and map to avoid the map lookup.

Reply:
I mean that, since we know ahead of time which characters belong in the `Map`, we can just construct the map with those character codes rather than looping and comparing. It's not that important, and if it makes more sense to have them visually grouped together, that's perfectly fine.